File: backend/app/schemas/workspace.py
"""Workspace / RBAC / RLS / audit schemas."""
from __future__ import annotations

from datetime import datetime
from typing import Any, Literal

from pydantic import BaseModel, Field

Role = Literal["viewer", "editor", "owner"]


class WorkspaceCreate(BaseModel):
    name: str = Field(min_length=1, max_length=255)


class WorkspaceResponse(BaseModel):
    id: str
    name: str
    owner_id: str
    role: str | None = None
    created_at: datetime

    model_config = {"from_attributes": True}


class MemberAdd(BaseModel):
    email: str = Field(min_length=3, max_length=255)
    role: Role = "viewer"


class MemberResponse(BaseModel):
    id: str
    user_id: str
    email: str
    role: str


class WorkspaceRename(BaseModel):
    name: str = Field(min_length=1, max_length=255)


class MemberRoleUpdate(BaseModel):
    # Promotion to owner goes through the transfer endpoint, not here.
    role: Literal["viewer", "editor"]


class TransferOwnership(BaseModel):
    member_id: str = Field(min_length=1)


class RLSRuleCreate(BaseModel):
    member_email: str = Field(min_length=3, max_length=255)
    column: str = Field(min_length=1, max_length=255)
    allowed_value: str = Field(min_length=1, max_length=500)


class RLSRuleResponse(BaseModel):
    id: str
    datasource_id: str
    member_id: str
    column: str
    allowed_value: str
    created_at: datetime

    model_config = {"from_attributes": True}


class AuditEntry(BaseModel):
    id: str
    action: str
    entity: str
    entity_id: str | None
    meta: dict[str, Any] | None
    created_at: datetime

    model_config = {"from_attributes": True}

File: frontend/src/api/workspace.ts
import { client } from './client'

export interface Workspace {
  id: string
  name: string
  owner_id: string
  role: string | null
  created_at: string
}

export interface WorkspaceMember {
  id: string
  user_id: string
  email: string
  role: string
}

export interface AuditEntry {
  id: string
  action: string
  entity: string
  entity_id: string | null
  meta: Record<string, unknown> | null
  created_at: string
}

export async function listWorkspaces(): Promise<Workspace[]> {
  const { data } = await client.get<Workspace[]>('/workspaces')
  return data
}

export async function createWorkspace(name: string): Promise<Workspace> {
  const { data } = await client.post<Workspace>('/workspaces', { name })
  return data
}

export async function deleteWorkspace(id: string): Promise<void> {
  await client.delete(`/workspaces/${id}`)
}

export async function listMembers(id: string): Promise<WorkspaceMember[]> {
  const { data } = await client.get<WorkspaceMember[]>(`/workspaces/${id}/members`)
  return data
}

export async function addMember(
  id: string,
  email: string,
  role: string,
): Promise<WorkspaceMember> {
  const { data } = await client.post<WorkspaceMember>(`/workspaces/${id}/members`, { email, role })
  return data
}

export async function removeMember(id: string, memberId: string): Promise<void> {
  await client.delete(`/workspaces/${id}/members/${memberId}`)
}

export async function renameWorkspace(id: string, name: string): Promise<Workspace> {
  const { data } = await client.patch<Workspace>(`/workspaces/${id}`, { name })
  return data
}

export async function changeMemberRole(
  id: string,
  memberId: string,
  role: string,
): Promise<WorkspaceMember> {
  const { data } = await client.patch<WorkspaceMember>(
    `/workspaces/${id}/members/${memberId}`,
    { role },
  )
  return data
}

export async function transferOwnership(id: string, memberId: string): Promise<void> {
  await client.post(`/workspaces/${id}/transfer`, { member_id: memberId })
}

export async function leaveWorkspace(id: string): Promise<void> {
  await client.post(`/workspaces/${id}/leave`)
}

export async function listAudit(): Promise<AuditEntry[]> {
  const { data } = await client.get<AuditEntry[]>('/audit')
  return data
}
